qualcommax: ipq807x: add support for Linksys MX5300

testuser7 opened the pull request:
Hardware specification:
========
SoC: Qualcomm IPQ8072A
Flash: 512MB (Winbond W29N04GZBIBA)
RAM: 1GB (2x Nanya DDR3L NT5CC256M16ER-EK)
Ethernet: 5x 10/100/1000Mbps (Qualcomm QCA8075)
WiFi1: 5GHz ac 4x4 (Qualcomm QCA9984 + Skyworks SKY85746-11) - channels 100-169
WiFi2: 5GHz ax 4x4 (Qualcomm QCN5054 + Skyworks SKY85755-11) - channels 36-64
WiFi3: 2.4GHz ax 4x4 (Qualcomm QCN5024 + Skyworks SKY8340-11)
IoT: Bluetooth 5, Zigbee and Thread (Qualcomm QCA4024 + Skyworks SE2433T-R)
IoT Flash: 4MB (Macronix MX25R3235F)
RTC: ST M41T00S
LED: 1x RGB status (NXP PCA9633)
USB: 1x USB 3.0
Button: WPS, Reset

Flash instructions:
========
1. Manually upgrade firmware using openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin image. More details can be found here: https://www.linksys.com/hk/support-article/?articleNum=274497
  After first boot check actual partition:
    - `fw_printenv -n boot_part`
  and install firmware on second partition using command in case of 2:
    - `mtd -r -e kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin kernel`
  and in case of 1:
    - `mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin alt_kernel`

2. Installation using serial connection from OEM firmware (default login: root, password: admin):
    - `fw_printenv -n boot_part`
  In case of 2:
    - `flash_erase /dev/mtd21 0 0`
      `nandwrite -p /dev/mtd21 openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin`
  or in case of 1:
    - `flash_erase /dev/mtd23 0 0`
      `nandwrite -p /dev/mtd23 openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin`
  After first boot install firmware on second partition:
    - `mtd -r -e kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin kernel`
  or:
    - `mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin alt_kernel`

3. Installation from initramfs image using USB FAT32 formatted drive:
  Stop u-boot and run:
    - `usb start && fatload usb 0:1 $loadaddr openwrt-qualcommax-ipq807x-linksys_mx5300-initramfs-uImage.itb && bootm $loadaddr`
  Write firmware to the flash from initramfs:
    - `mtd -e kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin kernel`
  and:
    - `mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin alt_kernel`

4. Back to the OEM firmware:
    - `mtd -e kernel -n write FW_MX5300_1.1.9.200251_prod.img kernel`
  and:
    - `mtd -r -e alt_kernel -n write FW_MX5300_1.1.9.200251_prod.img alt_kernel`

5. USB recovery using USB FAT32 formatted drive:
    - `fw_setenv usbimage 'openwrt-qualcommax-ipq807x-linksys_mx5300-initramfs-uImage.itb'`
      `fw_setenv bootusb 'usb start && fatload usb 0:1 $loadaddr $usbimage && bootm $loadaddr'`
      `fw_setenv bootcmd 'run bootusb; aq_load_fw && if test $auto_recovery = no; then bootipq; elif test $boot_part = 1; then run bootpart1; else run bootpart2; fi'`

Notes:
========
IoT device is accesible over spi. Not yet supported.

testuser7 commented:
Before merge this patch need to be upstreamed:
```
--- a/arch/arm64/boot/dts/qcom/ipq8074.dtsi
+++ b/arch/arm64/boot/dts/qcom/ipq8074.dtsi
@@ -572,6 +572,13 @@
 				bias-disable;
 			};
 
+			spi_3_pins: spi-3-state {
+				pins = "gpio50", "gpio52", "gpio53";
+				function = "blsp3_spi";
+				drive-strength = <8>;
+				bias-disable;
+			};
+
 			hsuart_pins: hsuart-state {
 				pins = "gpio46", "gpio47", "gpio48", "gpio49";
 				function = "blsp2_uart";
@@ -750,6 +757,23 @@
 			status = "disabled";
 		};
 
+		blsp1_spi4: spi@78b8000 {
+			compatible = "qcom,spi-qup-v2.2.1";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x078b8000 0x600>;
+			interrupts = <GIC_SPI 98 IRQ_TYPE_LEVEL_HIGH>;
+			spi-max-frequency = <50000000>;
+			clocks = <&gcc GCC_BLSP1_QUP4_SPI_APPS_CLK>,
+				<&gcc GCC_BLSP1_AHB_CLK>;
+			clock-names = "core", "iface";
+			dmas = <&blsp_dma 18>, <&blsp_dma 19>;
+			dma-names = "tx", "rx";
+			pinctrl-0 = <&spi_3_pins>;
+			pinctrl-names = "default";
+			status = "disabled";
+		};
+
 		blsp1_i2c5: i2c@78b9000 {
 			compatible = "qcom,i2c-qup-v2.2.1";
 			#address-cells = <1>;
```

BDF's are not committed yet.

kirdesde commented:
Please do not use mtd to write to NAND partitions. mtd doesn't take care of bad block management and can damage the NAND on the device.
Use nand tools instead. Please rework your install instructions accordingly.

kirdesde commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,555 @@
+
+	nand@0 {
+		reg = <0>;
+		compatible = "qcom,nandcs";
```
That compatible doesn't exists upstream, rm that property

testuser7 commented:
> Please do not use mtd to write to NAND partitions. mtd doesn't take care of bad block management and can damage the NAND on the device. Use nand tools instead. Please rework your install instructions accordingly.

mtd support bad block management: http://www.linux-mtd.infradead.org/tech/mtdnand/x144.html

kirdesde commented:
> > Please do not use mtd to write to NAND partitions. mtd doesn't take care of bad block management and can damage the NAND on the device. Use nand tools instead. Please rework your install instructions accordingly.
> 
> mtd support bad block management: http://www.linux-mtd.infradead.org/tech/mtdnand/x144.html

Yes MTD devices itself supports bad block management, but not the mtd tool we are using in OpenWrt.

kirdesde commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -33,6 +33,7 @@ ALLWIFIBOARDS:= \
 	dynalink_dl-wrx36 \
 	edgecore_eap102 \
 	edimax_cax1800 \
+  linksys_mx5300 \
```
wrong indent

kirdesde commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,555 @@
+&switch {
+	status = "okay";
+
+	switch_cpu_bmp = <0x1>; /* cpu port bitmap */
```
That needs to updated to the MAC mode defines:

https://github.com/openwrt/openwrt/commit/cb1702041f8ccfa902b6e525b2268313aeaf1eb0
and
https://github.com/openwrt/openwrt/commit/f3cd4bfb7f697bb885e976627ac2ecbb99e80c60

kirdesde commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,555 @@
+	switch_mac_mode = <0x0>; /* mac mode for uniphy instance0*/
+	switch_mac_mode1 = <0xff>; /* mac mode for uniphy instance1*/
+	switch_mac_mode2 = <0xd>; /* mac mode for uniphy instance2*/
+	bm_tick_mode = <0>; /* bm tick mode */
```
tick mode has been also moved to the ESS dtsi
https://github.com/openwrt/openwrt/commit/c885169e02a56a889f706b7734599ac42dcf9886

kirdesde commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,555 @@
+	label = "wan";
+};
+
+&dp6_syn {
```
Disabling a non existing dp should be necessary any more since commit:

https://github.com/openwrt/openwrt/commit/e927456ec3d9cd7e92ba884a3065ecc303564720

rm port@5 under qcom,port_phyinfo as well and check again if there "phydev is NULL error message"  are really still there

kirdesde commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/firmware/11-ath11k-caldata:
```diff
@@ -27,15 +28,15 @@ case "$FIRMWARE" in
 		caldata_extract "0:art" 0x1000 0x20000
 		;;
 	prpl,haze)
-                caldata_extract_mmc "0:ART" 0x1000 0x20000
```
I guess this change is unintentionally?

testuser7 replied:
This is just an indent correction.

robimarko replied:
All of these need to go into a separate commit.

kirdesde commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/firmware/11-ath11k-caldata:
```diff
@@ -27,15 +28,15 @@ case "$FIRMWARE" in
 "ath11k/QCN9074/hw1.0/cal-pci-0001:01:00.0.bin")
 	case "$board" in
 	prpl,haze)
-                caldata_extract_mmc "0:ART" 0x26800 0x20000
```
I guess this change is unintentionally?

testuser7 replied:
This is just an indent correction.

robimarko replied:
Same here

robimarko commented:
> Before merge this patch need to be upstreamed:
> 
> ```
> --- a/arch/arm64/boot/dts/qcom/ipq8074.dtsi
> +++ b/arch/arm64/boot/dts/qcom/ipq8074.dtsi
> @@ -572,6 +572,13 @@
>  				bias-disable;
>  			};
>  
> +			spi_3_pins: spi-3-state {
> +				pins = "gpio50", "gpio52", "gpio53";
> +				function = "blsp3_spi";
> +				drive-strength = <8>;
> +				bias-disable;
> +			};
> +
>  			hsuart_pins: hsuart-state {
>  				pins = "gpio46", "gpio47", "gpio48", "gpio49";
>  				function = "blsp2_uart";
> @@ -750,6 +757,23 @@
>  			status = "disabled";
>  		};
>  
> +		blsp1_spi4: spi@78b8000 {
> +			compatible = "qcom,spi-qup-v2.2.1";
> +			#address-cells = <1>;
> +			#size-cells = <0>;
> +			reg = <0x078b8000 0x600>;
> +			interrupts = <GIC_SPI 98 IRQ_TYPE_LEVEL_HIGH>;
> +			spi-max-frequency = <50000000>;
> +			clocks = <&gcc GCC_BLSP1_QUP4_SPI_APPS_CLK>,
> +				<&gcc GCC_BLSP1_AHB_CLK>;
> +			clock-names = "core", "iface";
> +			dmas = <&blsp_dma 18>, <&blsp_dma 19>;
> +			dma-names = "tx", "rx";
> +			pinctrl-0 = <&spi_3_pins>;
> +			pinctrl-names = "default";
> +			status = "disabled";
> +		};
> +
>  		blsp1_i2c5: i2c@78b9000 {
>  			compatible = "qcom,i2c-qup-v2.2.1";
>  			#address-cells = <1>;
> ```
> 
> BDF's are not committed yet.

I have sent the node upstream:
https://patchwork.kernel.org/project/linux-arm-msm/patch/20231123121324.1046164-1-robimarko@gmail.com/

Feel free to include it as a separate commit in this PR.
Pins though you need to define in your board DTS as there are 2 sets of pins that can be used for this QUP SPI controller.

testuser7 commented:
> Yes MTD devices itself supports bad block management, but not the mtd tool we are using in OpenWrt.

In `nand.sh` script mtd is used to write kernel part: https://github.com/openwrt/openwrt/blob/main/package/base-files/files/lib/upgrade/nand.sh#L351
And here is the answer on the forum: https://forum.openwrt.org/t/nandwrite-vs-mtd-write/151033/2

testuser7 committed:
qualcommax: ipq807x: add support for Linksys MX5300

Hardware specification:
========
SoC: Qualcomm IPQ8072A
Flash: 512MB (Winbond W29N04GZBIBA)
RAM: 1GB (2x Nanya DDR3L NT5CC256M16ER-EK)
Ethernet: 5x 10/100/1000Mbps (Qualcomm QCA8075)
WiFi1: 5GHz ac 4x4 (Qualcomm QCA9984 + Skyworks SKY85746-11) - channels 100-169
WiFi2: 5GHz ax 4x4 (Qualcomm QCN5054 + Skyworks SKY85755-11) - channels 36-64
WiFi3: 2.4GHz ax 4x4 (Qualcomm QCN5024 + Skyworks SKY8340-11)
IoT: Bluetooth 5, Zigbee and Thread (Qualcomm QCA4024 + Skyworks SE2433T-R)
IoT Flash: 4MB (Macronix MX25R3235F)
RTC: ST M41T00S
LED: 1x RGB status (NXP PCA9633)
USB: 1x USB 3.0
Button: WPS, Reset

Flash instructions:
========
1. Manually upgrade firmware using openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin image.
More details can be found here: https://www.linksys.com/hk/support-article?articleNum=274497
After first boot check actual partition:
- fw_printenv -n boot_part
and install firmware on second partition using command in case of 2:
- mtd -r -e kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin kernel
and in case of 1:
- mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin alt_kernel

2. Installation using serial connection from OEM firmware (default login: root, password: admin):
- fw_printenv -n boot_part
In case of 2:
- flash_erase /dev/mtd21 0 0
  nandwrite -p /dev/mtd21 openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin
or in case of 1:
- flash_erase /dev/mtd23 0 0
  nandwrite -p /dev/mtd23 openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin
After first boot install firmware on second partition:
- mtd -r -e kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin kernel
or:
- mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin alt_kernel

3. Installation from initramfs image using USB FAT32 formatted drive:
Stop u-boot and run:
- usb start && fatload usb 0:1 $loadaddr openwrt-qualcommax-ipq807x-linksys_mx5300-initramfs-uImage.itb && bootm $loadaddr
Write firmware to the flash from initramfs:
- mtd -e kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin kernel
and:
- mtd -r -e alt_kernel -n write openwrt-qualcommax-ipq807x-linksys_mx5300-squashfs-factory.bin alt_kernel

4. Back to the OEM firmware:
- mtd -e kernel -n write FW_MX5300_1.1.9.200251_prod.img kernel
and:
- mtd -r -e alt_kernel -n write FW_MX5300_1.1.9.200251_prod.img alt_kernel

5. USB recovery:
- fw_setenv usbimage 'openwrt-qualcommax-ipq807x-linksys_mx5300-initramfs-uImage.itb'
  fw_setenv bootusb 'usb start && fatload usb 0:1 $loadaddr $usbimage && bootm $loadaddr'
  fw_setenv bootcmd 'run bootusb; aq_load_fw && if test $auto_recovery = no; then bootipq; elif test $boot_part = 1; then run bootpart1; else run bootpart2; fi'

Notes:
========
IoT device is accesible over spi. Not yet supported.

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>
Reviewed-by: Robert Marko <robimarko@gmail.com>

testuser7 reviewed (commented):
DTS and indent corrected

robimarko committed:
qualcommax: ipq807x: add patch for QUP4 SPI node

Add node to support the QUP4 SPI controller inside of IPQ8074.
Some devices use this bus to communicate to a Bluetooth controller.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -48,16 +48,16 @@ endef
 TARGET_DEVICES += buffalo_wxr-5950ax12
 
 define Device/compex_wpq873
-       $(call Device/FitImage)
```
Please all of the indent fixes, reordering etc into separate commits.

testuser7 replied:
Is it okay to add these commits to this PR?

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+		};
+	};
+
+	spi_3_pins: spi-3-state {
```
Are you sure that this gets properly applied?

Because, this is completely against the bindings the way its currently structured.

testuser7 replied:
I'm not sure how pins should be described. Here is the OEM dts:
- https://gist.github.com/testuser7/3db5a8ff8b621ae1a216dd919ce2113a#file-mx5300-dts-L3134-L3155
- https://gist.github.com/testuser7/3db5a8ff8b621ae1a216dd919ce2113a#file-mx5300-dts-L2075-L2106

robimarko replied:
I dont care about vendor DTS-es, they are usually broken in so many ways.

They should be like:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/arch/arm64/boot/dts/qcom/ipq8074.dtsi?h=next-20231128#n343

Just change the naming and pins to the correct ones for this QUP.

Those Quartz pins should be a separate node that you can just include along spi3_pins in the spidev node

testuser7 replied:
It's not obvious to me how the pins should be configured.
Is this correct?
```
	spi_3_default: spi-3-default {
		mux {
			pins = "gpio50", "gpio52", "gpio53";
			function = "blsp3_spi";
			drive-strength = <8>;
			bias-disable;
		}

		spi_cs {
			pins = "gpio22";
			function = "blsp3_spi2";
			drive-strength = <8>;
			bias-disable;
		};
	};

	spi_3_interrupt: spi-3-interrupt {
		pins = "gpio48";
		function = "gpio";
		bias-disable;
		input;
	};

	spi_3_reset: spi-3-reset {
		pins = "gpio21";
		function = "gpio";
		bias-disable;
		output-low;
	};
   
&blsp1_spi4 {
	pinctrl-names = "default", "interrupt", "reset";
	pinctrl-0 = <&spi_3_default>;
	pinctrl-1 = <&spi_3_interrupt>;
	pinctrl-2 = <&spi_3_reset>;
	...
```
or maybe this way?
```
	spi_3_pins: spi-3-state {
		mux {
			pins = "gpio50", "gpio52", "gpio53";
			function = "blsp3_spi";
			drive-strength = <8>;
			bias-disable;
		}

		quartz_interrupt {
			pins = "gpio48";
			function = "gpio";
			bias-disable;
			input;
		};

		quartz_reset {
			pins = "gpio21";
			function = "gpio";
			bias-disable;
			output-low;
		};

		spi_cs {
			pins = "gpio22";
			function = "blsp3_spi2";
			drive-strength = <8>;
			bias-disable;
		};
	};
	
&blsp1_spi4 {
	pinctrl-names = "default";
	pinctrl-0 = <&spi_3_pins>;
	...
```

robimarko replied:
```
&tlmm {
	spi_3_pins: spi-3-state {
		spi-pins {
			pins = "gpio50", "gpio52", "gpio53";
			function = "blsp3_spi";
			drive-strength = <8>;
			bias-disable;
		};

		cs-pins {
			pins = "gpio22";
			function = "blsp3_spi2";
			drive-strength = <8>;
			bias-disable;
		};
	};

	quartz_pins: quartz-state {
		interrupt-pins {
			pins = "gpio48";
			function = "gpio";
			bias-disable;
			input;

		};

		reset-pins {
			pins = "gpio21";
			function = "gpio";
			bias-disable;
			output-low;
		};
	};
};

&blsp1_spi4 {
	pinctrl-0 = <&spi_3_pins &quartz_pins>;
	pinctrl-names = "default";
};
```

This should work while being mostly compliant to DT bindings

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+		drive-strength = <8>;
+		bias-disable;
+
+		quartz_interrupt {
```
These are Bluetooth?

If so, then it should be a completely separate pinctrl node.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+
+	rtc@68 {
+		compatible = "dallas,ds1307";
+		reg = <104>;
```
I2C always uses hex reg properties.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+		compatible = "nxp,pca9633";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <98>;
```
I2C always uses hex reg properties.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+	};
+};
+
+&blsp1_spi1 {
```
Is there anything connected to this SPI bus?

testuser7 replied:
I'm not sure. There is a 4MB IoT flash memory (Macronix MX25R3235F), but I don't know how it is connected.

robimarko replied:
Why even enable this bus then?

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+&blsp1_spi4 {
+	pinctrl-0 = <&spi_3_pins>;
+	pinctrl-names = "default";
+	cs-gpios = <2>;
```
This cannot be right

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+	cs-gpios = <2>;
+	status = "okay";
+
+	spidev1: spi@3 {
```
Do you need the label at all?

Also, its definitively not `spi@3`, I assume this is Bluetooth controller so probably `bluetooth@0`

testuser7 replied:
This is multi-mode SoC QCA4024, Bluetooth and 802.15.4 (Zigbee, Thread).

robimarko replied:
Then for the lack of better ideas `qca4024@0`.

testuser7 replied:
Or maybe `iot@0`?

robimarko replied:
`iot` is fine for me, just leave a comment with the actual device model.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+
+	spidev1: spi@3 {
+		/*
+		 * Compatible is changed from "qca,spidev"
```
This comment is not needed

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+	switch_lan_bmp = <(ESS_PORT1 | ESS_PORT2 | ESS_PORT3 | ESS_PORT4)>; /* lan port bitmap */
+	switch_wan_bmp = <ESS_PORT5>; /* wan port bitmap */
+	switch_mac_mode = <MAC_MODE_PSGMII>; /* mac mode for uniphy instance0*/
+	switch_mac_mode2 = <MAC_MODE_USXGMII>; /* mac mode for uniphy instance2*/
```
This cannot be right as DP6 is unused and no PHY is defined

testuser7 replied:
So `switch_mac_mode2` should be removed?

robimarko replied:
Yes, just remove it as its disabled by default

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,539 @@
+		 * Compatible is changed from "qca,spidev"
+		 * to be able to load driver
+		 */
+		compatible = "silabs,si3210";
```
Since this is QCA4024 this compatible for sure isn't proper

testuser7 replied:
Qualcomm was using `compatible = "spidev"`. But now this is not possible: https://github.com/torvalds/linux/blob/v6.1/drivers/spi/spidev.c#L704-L708
They expand the compatibility list: https://github.com/dissent1/msm-2/commit/d6160218393552fea1b7973787f2bd154f870ee2
Should a similar patch be added? As in the case of `si3210`: https://github.com/openwrt/openwrt/blob/main/target/linux/generic/pending-6.1/110-v6.3-0002-spidev-Add-Silicon-Labs-SI3210-device-compatible.patch

robimarko replied:
Yes, a new compatible should be added for your device upstream.

testuser7 replied:
@robimarko: pings on this.

robimarko replied:
Not really, I dont have time currently

testuser7 replied:
If anyone can upstream this patch, I would appreciate it.

robimarko replied:
You can do it yourself, it shouldnt be hard

testuser7 committed:
qualcommax: ipq807x: reordering, combine the same commands

Reordering, combine the same commands for ipq807x devices

Signed-off-by: Paweł Owoc <frut3k7@gmail.com>

raenye commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,531 @@
+		nxp,hw-blink;
+
+		led_system_red: led@0 {
+			label = "red";
```
```suggestion
```
Label is deprecated (but you could add a function, e.g. `function = LED_FUNCTION_STATUS`).

raenye commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,531 @@
+		};
+
+		led_system_green: led@1 {
+			label = "green";
```
```suggestion
```
Label is deprecated (but you could add a function, e.g. `function = LED_FUNCTION_STATUS`).

raenye commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,531 @@
+		};
+
+		led_system_blue: led@3 {
+			label = "blue";
```
```suggestion
```
Label is deprecated (but you could add a function, e.g. `function = LED_FUNCTION_STATUS`).

testuser7 commented:
Is there any way to disable the creation of the ath10k led in the system?
```
root@OpenWrt:~# ls -l /sys/class/leds/
lrwxrwxrwx    1 root     root             0 Dec  8 13:45 ath10k-phy0 -> ../../devices/platform/soc/10000000.pci/pci0001:00/0001:00:00.0/0001:01:00.0/leds/ath10k-phy0
```

raenye commented:
> Is there any way to disable the creation of the ath10k led in the system?

Why would you?

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -101,6 +101,26 @@ define Device/edimax_cax1800
+	KERNEL_SIZE := 6144k
+	IMAGE_SIZE := 147456k
+	NAND_SIZE := 512m
+	KERNEL_IN_UBI :=
```
Can you clarify?

testuser7 commented:
Is it possible to change the order of radio interfaces detected in the system?
Currently, the interface on pcie is detected as `phy0`. Can it be detected as `phy2`?

raenye commented:
> Is it possible to change the order

I couldn't find how to do it (on a different platform).
(Tried changing node order in the DTS)

testuser7 commented:
For the IoT chip (QCA4024) to make it work some initialization in the OEM firmware is done:
```
#move to here for Bronx, because Quartz need more time to become ready
# Get BTLE radio ready
echo "[utopia][init] Initializing BTLE interface" >> /dev/console
BTRESETGPIO="21"
echo $BTRESETGPIO >/sys/class/gpio/export
echo out >/sys/class/gpio/gpio${BTRESETGPIO}/direction
#moved to iotd
#echo 0 >/sys/class/gpio/gpio${BTRESETGPIO}/value
#sleep 0.05
#echo 1 >/sys/class/gpio/gpio${BTRESETGPIO}/value
#sleep 0.05

# Program BT Mac address
BT_MAC_ADDR=`syscfg get bt_mac_addr`
BT_MAC_ADDR=`echo "$BT_MAC_ADDR" | tr -s ':' ' '`
MAC0=`echo "$BT_MAC_ADDR" | awk '{print $1}'`
MAC1=`echo "$BT_MAC_ADDR" | awk '{print $2}'`
MAC2=`echo "$BT_MAC_ADDR" | awk '{print $3}'`
MAC3=`echo "$BT_MAC_ADDR" | awk '{print $4}'`
MAC4=`echo "$BT_MAC_ADDR" | awk '{print $5}'`
MAC5=`echo "$BT_MAC_ADDR" | awk '{print $6}'`
CERT_REGION=`skuapi -g cert_region | cut -d'=' -f2 | tr -d ' '`
# Dallas uses the same psr file for all cert regions.
#PSR="/etc/csr8x11-a12-bt4.2.psr"

echo "[utopia][init] Bluetooth Mac Address: $BT_MAC_ADDR" >> /dev/console
echo "[utopia][init] Cert Region: $CERT_REGION" >> /dev/console
#echo "[utopia][init] BT PSR: $PSR" >> /dev/console

if [ -c /dev/spidev32765.0 ] ; then
	echo "[utopia][init] BTLE interface ready to use" >> /dev/console
    # Start btsetup service
    sysevent set btsetup-start
    sysevent set quartzfwup 0
else
	echo "[utopia][init] BTLE interface failed" >> /dev/console
fi
```
Should such initialization be done in OpenWrt by default?

raenye commented:
> Should such initialization be done in OpenWrt by default?

I think it should; certainly some (perhaps all) of it can be done via the DTS.
It seems that besides pulsing the reset GPIO, the rest of the script just sets up environment variables and calls userspace tools.

testuser7 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,532 @@
+			pins = "gpio21";
+			function = "gpio";
+			bias-disable;
+			output-high;
```
Is it correct to set the high state this way?

robimarko replied:
According to bindings that is valid

testuser7 commented:
MAC address for wired interfaces is stored in two different places:
- uboot-env
- `devinfo` partition

Where should we read the MAC address from?

Some Linksys devices use `devinfo` partition: https://github.com/openwrt/openwrt/blob/main/target/linux/ipq806x/base-files/etc/board.d/02_network#L79-L88

robimarko commented:
> MAC address for wired interfaces is stored in two different places:
> 
> * uboot-env
> * `devinfo` partition
> 
> Where should we read the MAC address from?
> 
> Some Linksys devices use `devinfo` partition: https://github.com/openwrt/openwrt/blob/main/target/linux/ipq806x/base-files/etc/board.d/02_network#L79-L88

I would prefer to read from `devinfo` directly, U-Boot most likely reads from there anyway.

testuser7 commented:
> I would prefer to read from `devinfo` directly, U-Boot most likely reads from there anyway.

Is it possible to read the MAC address as ASCII text from DTS?
`devinfo` partition:
```
root@OpenWrt:~# hexdump -C /dev/mtd20
00000000  6d 66 67 5f 64 61 74 61  5f 76 65 72 73 69 6f 6e  |mfg_data_version|
00000010  3d 31 0a 6d 6f 64 65 6c  4e 75 6d 62 65 72 3d 4d  |=1.modelNumber=M|
00000020  58 35 33 30 30 2d 45 55  0a 73 65 72 69 61 6c 5f  |X5300-EU.serial_|
...
```

robimarko commented:
Yes, you can read it via NVMEM, check
https://github.com/openwrt/openwrt/commit/931fcf61891f14dffccd0ad5dfdf7d84460ff9f2

Or plenty other examples

testuser7 commented:
> Yes, you can read it via NVMEM, check [931fcf6](https://github.com/openwrt/openwrt/commit/931fcf61891f14dffccd0ad5dfdf7d84460ff9f2)
> 
> Or plenty other examples

But will the MAC address always be at the same address?
I was thinking of reading key and value like in here: https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts#L217-L225

robimarko commented:
Oh yeah, I was on my phone and completely missed that it resembles to an U-Boot environment.

You can try using the U-Boot parser, it should work

testuser7 commented:
> Oh yeah, I was on my phone and completely missed that it resembles to an U-Boot environment.
> 
> You can try using the U-Boot parser, it should work

Unfortunately it is not possible to use the u-boot parser. `ethaddr` is a property, not a variable name.
It's possible to read the MAC address as ASCII but only from a specific address: https://github.com/openwrt/openwrt/blob/main/target/linux/ath79/dts/ar9342_zyxel_nwa11xx.dtsi#L95-L111
And this address is not permanent...

testuser7 commented on target/linux/qualcommax/ipq807x/base-files/etc/board.d/02_network:
```diff
@@ -55,9 +54,32 @@ ipq807x_setup_interfaces()
+		linksys,mx5300)
+			label_mac=$(mtd_get_mac_ascii devinfo hw_mac_addr)
+			for i in $(seq 1 4); do
+				[ "$(mtd_get_mac_ascii u_env eth${i}addr)" != "$label_mac" ] && lan_mac=$label_mac
```
@robimarko What do you think about this implementation of MAC address setting as setting from DTS is not possible?

robimarko replied:
Why is it not possible?

Does U-Boot env have `ethaddr` variable?
If so, you can then just increment all other MAC-s via it.

testuser7 replied:
On some devices `ethaddr`, `eth1addr`, ... are different than `hw_mac_addr` from `devinfo` partition.

robimarko replied:
Ugh, I remember now that the U-Boot parser is not yet converted to NVMEM layout and it has the `ethaddr` variable hardcoded for MAC, so you cannot parse custom variable names.

If U-Boot sees different MAC-s, I suppose it then uses the "wrong" MAC-s for patching the DTS as well?

testuser7 replied:
> If U-Boot sees different MAC-s, I suppose it then uses the "wrong" MAC-s for patching the DTS as well?

Yes, MAC addresses are incorrect.
For example Linksys MX42CF (branded MX4200) has:
- `hw_mac_addr`: `80:69:1A:xx:xx:xx`
- `ethaddr`: `E8:9F:80:xx:xx:xx`
- `caldata`: `00:03:7F:xx:xx:xx`

testuser7 replied:
For my router, the MAC addresses are correct in all three places. Is it worth implementing MAC address setting? I don't know if there is any branded version of the MX5300.

robimarko replied:
I am not following, if they are different in those 2 places, which one is correct?

testuser7 replied:
Correct one is `hw_mac_addr` from `devinfo` partition.

testuser7 replied:
I removed the MAC address changes completely. There is currently no known MX5300 with incorrectly configured addresses.

testuser7 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-mx5300.dts:
```diff
@@ -0,0 +1,532 @@
+
+			qcom,ath10k-calibration-variant = "Linksys-MX5300";
+			nvmem-cell-names = "calibration";
+			nvmem-cells = <&caldata_qca9984>;
```
If we set calibration data here, do we no longer need to use `caldata_extract` function?

robimarko replied:
No, but in caldata_extract you called this pre-cal and here its called calibration which would be equal to a file called caldata.

So, my question is which one is correct?
QCA9948 usually uses `pre-calibration`

testuser7 replied:
How can I check whether it is `calibration` or `pre-calibration`?

robimarko replied:
Yes

testuser7 commented on target/linux/qualcommax/ipq807x/base-files/etc/board.d/02_network:
```diff
@@ -55,9 +54,32 @@ ipq807x_setup_interfaces()
+
+	[ -n "$lan_mac" ] && ucidef_set_interface_macaddr "lan" $lan_mac
+	[ -n "$wan_mac" ] && ucidef_set_interface_macaddr "wan" $wan_mac
+	[ -n "$label_mac" ] && ucidef_set_label_macaddr $label_mac
```
Where should the label MAC address be set so that it can be used in hotplug.d scripts?
If we do it here, the `get_mac_label` function does not return the correct address in hotplug.d scripts during firstboot.

robimarko replied:
It can be set via 02_network or via an alias in DTS, depending on board

testuser7 replied:
Setting via DTS is not possible here.
The setting from `02_network` works correctly, but with `firstboot` the function `get_mac_label` called in the `hotplug.d` scripts returns `00:00:0*`.

What is the startup order?
1. init.d
2. hotplug.d
3. board.d

robimarko replied:
The hotplug script probably starts before 02_board

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/firmware/10-ath10k-caldata:
```diff
@@ -0,0 +1,23 @@
+	linksys,mx5300)
+		caldata_extract "0:art" 0x33000 0x2f20
+		ath10k_patch_mac $(macaddr_add $(get_mac_label) 3)
+		ln -sf /lib/firmware/ath10k/pre-cal-pci-0001\:01\:00.0.bin \
```
Sorry, but WTF is this?

Precal data is not BDF, I have seen this craziness being done before but its not valid.
Also, for ath10k you can load the pre-cal and caldata via NVMEM

testuser7 replied:
I know it's set up the wrong way at the moment.
`pre-calibration` is a data saved on the `art` partition and `calibration` is the BDF file included in the OEM firmware? Or maybe the other way around?
But if we want to change the MAC address outside DTS, this is probably how it should be done?

robimarko replied:
Sorry but no, this is completely wrong as you are pretending that pre-cal data is BDF, and it's a completely different thing.

Remove the symbolic link please.

testuser7 replied:
The file has been deleted.

robimarko commented:
Can you replace SPI patch with the upstreamed version:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/arch/arm64/boot/dts/qcom/ipq8074.dtsi?h=next-20240201&id=6a25e70214fde6dcf900271c819c8d7fe7b9a4b0

Also, please mark it under the `00xx-v6.8-patch-name` format.

After that:
`Reviewed-by: Robert Marko <robimarko@gmail.com>`

testuser7 commented:
> Can you replace SPI patch with the upstreamed version: https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/arch/arm64/boot/dts/qcom/ipq8074.dtsi?h=next-20240201&id=6a25e70214fde6dcf900271c819c8d7fe7b9a4b0
> 
> Also, please mark it under the `00xx-v6.8-patch-name` format.
> 
> After that: `Reviewed-by: Robert Marko <robimarko@gmail.com>`

Done, thanks!

I noticed that on every boot when mounting `rootfs_data` there is a recovery needed:
```
[    7.451257] UBIFS (ubi0:1): Mounting in unauthenticated mode
[    7.451360] UBIFS (ubi0:1): background thread "ubifs_bgt0_1" started, PID 814
[    7.489233] UBIFS (ubi0:1): recovery needed
[    7.582516] UBIFS (ubi0:1): recovery completed
[    7.582608] UBIFS (ubi0:1): UBIFS: mounted UBI device 0, volume 1, name "rootfs_data"
[    7.585876] UBIFS (ubi0:1): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
[    7.593761] UBIFS (ubi0:1): FS size: 126595072 bytes (120 MiB, 997 LEBs), max 1008 LEBs, journal size 6348800 bytes (6 MiB, 50 LEBs)
[    7.603669] UBIFS (ubi0:1): reserved for root: 4952683 bytes (4836 KiB)
[    7.615637] UBIFS (ubi0:1): media format: w5/r0 (latest is w5/r0), UUID *uuid*, small LPT model

```
Does something need to be done about this?

robimarko commented:
Can you please rebase this, I will give it another pass and if all is fine merge it.

Can you also tell me if BDF has been merged?

testuser7 commented:
> Can you please rebase this, I will give it another pass and if all is fine merge it.

Done.

> Can you also tell me if BDF has been merged?

Not yet. PR is ready: https://github.com/openwrt/firmware_qca-wireless/pull/23

robimarko commented:
Thanks,
What was the conclusion of upstream discussion about the SPI device compatible?

testuser7 commented:
> Thanks, What was the conclusion of upstream discussion about the SPI device compatible?

No response yet (after compatibility name change) from the spi project.

However, the reaction of the devicetree project is rather negative. They want to describe all the interfaces that can be used to communicate with the QCA4024.

robimarko commented:
I get their stance, so in order not to further delay this can you remove the QCA4024 patch and add a comment to the DTS that its not yet supported

testuser7 commented:
> I get their stance, so in order not to further delay this can you remove the QCA4024 patch and add a comment to the DTS that its not yet supported

Maybe add this patch as a hack? Support for QCA4024 will likely never be added upstream.

robimarko commented:
No, keeping it as a hack doesnt make it better, it just means that OpenWrt will have to maintain it downstream forever which is not what anybody wants.

Also, please stop poking at the upstream maintainers with top-posting etc as that will just get them to ignore you.

testuser7 commented:
> No, keeping it as a hack doesnt make it better, it just means that OpenWrt will have to maintain it downstream forever which is not what anybody wants.

Do you want to completely remove the device from DTS or just the patch?

robimarko commented:
> Do you want to completely remove the device from DTS or just the patch?

You can remove the patch, and then comment out that part of the DTS with a comment that its for documentation purposes as there is no SW support currently.

testuser7 commented:
> I looked at the specification sheet for QCA4024 and they are right, it is way more complicated than a trivial compatible, its a fully fleghed standalone SoC

But Linksys uses it as trivial spidev.

robimarko commented:
What ODM-s do is completely irrelevant

testuser7 commented:
> You can remove the patch, and then comment out that part of the DTS with a comment that its for documentation purposes as there is no SW support currently.

Done.

> Does it just work via bluez if exposed as spidev or is there additional userspace SW used?

BlueZ does not support devices over spi. If connected via UART it should work.

In the case of QCA4024 connected via spi, communication takes place via the QAPI protocol.

I'm not really interested in Bluetooth. I was thinking about using Thread radio: https://github.com/openwrt/packages/pull/19218

robimarko commented:
BDF has been merged, so make sure to update `ipq-wifi`

testuser7 commented:
> BDF has been merged, so make sure to update `ipq-wifi`

Updated.

robimarko commented:
Thanks! Rebased on top of main and merged!
